src/patterns.jl

```diff
@@ -0,0 +1,82 @@
+
+"""
+    Patter
+
+A pattern is just a wrapper around a `Tuple{Vararg{Symbol}}` which enables searching for
+matching components and dimension paths in a Dataset.
+Special symbols `:_` and `:__` are used as wildcards, similar to `*` and `**` in glob
+pattern matching.
+"""
+struct Pattern
+    segments::Tuple{Vararg{Symbol}}
+end
+
+Pattern(segments::Symbol...) = Pattern(segments)
+Base.convert(::Type{Pattern}, x::Tuple{Vararg{Symbol}}) = Pattern(x)
+
+# Primary functionality is to support identify tuples that match the constraint
+function Base.in(item::Tuple{Vararg{Symbol}}, pattern::Pattern)
+    pat = pattern.segments
+    i = 1   # item  index
+    j = 1   # contraint index
+    n = length(item)
+    m = length(pat)
+
+    # If our constrain has more segments than the item then it isn't going to match
+    m <= n || return false
+
+    result = true
+    while result && i <= n && j <= m
```

Comment:
Would it be clearer to just convert `pattern` to a regex expression and try to match it against a stringified `item`?

Reply:
The internal code might be shorter/clearer, but it'll likely be slower to stringify and construct the regex on every call to `in`.